tools/audioconv64/conv_ym64.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * At runtime, ym64.c supports streaming directly from a YM5/YM6
+ * "interleaved" file, that is a file where frames are laid out linearly in the
+ * file. We also support streaming from LHA-compressed files (using algorithm
+ * -lh5-, which is the standard ones for YM files).
+ *
+ * The goal of this pre-processing is:
+ *
+ *   * Convert from older YM versions (eg: YM3!, YM3b).
+ *   * Convert non-interleaved to interleaved.
+ *   * Re-compress with LHA -lh5-.
+ *
+ */
+
+#include "../../src/audio/lzh5.h"   // LZH5 decompression
+#include "lzh5_compress.h"          // LZH5 compression
+#include "lzh5_compress.c"
+
+bool flag_ym_compress = false;
+
+typedef struct __attribute__((packed)) {
+    uint8_t size;
+    uint8_t checksum;
+    char methodid[5];
+    uint32_t csize;
+    uint32_t dsize;
+    uint32_t timestamp;
+    uint8_t attr;
+    uint8_t level;
+    uint8_t filename_len;
+} lhaheader;
+
+_Static_assert(sizeof(lhaheader) == 22, "invalid lhaheader size");
+
+typedef struct __attribute__((packed)) {
+    uint32_t nvbl;
+    uint32_t attrs;
+    uint16_t ndigidrums;
+    uint32_t extfreq;
+    uint16_t playfreq;
+    uint32_t loop;
+    uint16_t sizeext;
+} ym5header;
+
+_Static_assert(sizeof(ym5header) == 22, "invalid ym5header size");
+
+static FILE *ym_f;
+static bool ym_compressed;
+static LHANewDecoder ym_decoder;
+
+static size_t lha_callback(void *buf, size_t buf_len, void *user_data) {
+    return fread(buf, 1, buf_len, ym_f);
+}
+
+static void ymread(void *buf, int sz) {
+    if (ym_compressed) {
+        lha_lh_new_read(&ym_decoder, buf, sz);
+        return;
+    }
+    fread(buf, 1, sz, ym_f);
+}
+
+static void ymwrite(const void *buf, int sz) {
+    fwrite(buf, 1, sz, ym_f);
+}
+
+// Compress a file with LHA (algorithm -lh5-). This is done using
+// a stripped down version of https://github.com/jca02266/lha
+// stored as single file in lzh5_compress.c. The library works only
+// through FILE*, so the compression will happen from disk to disk.
```

Comment:
It looks like it won't work. I have two codepaths: 1) no compression: write to final file directly. 2) compression: write to temporary file; then compress from temporary to final file. `open_memstream` doesn't allow to read from it, only write (duh!), so It can't replace the temporary file.